### PR TITLE
Correct handling of weight decay in Adam and RMSProp, uncentered version of RMSProp, weight clipping for RMSProp

### PR DESCRIPTION
**Pull request by leezu**
See #5099 on the handling of weight decay.

Weight clipping is necessary to train WassersteinGAN. Clipping the weights directly in the optimizer speeds things up compared to iterating over the weight / bias arrays in Python and performing the weight clipping there. WassersteinGAN are typically trained with RMSProp.

This pull request obsoletes #5116 and introduces the uncentered version of RMSProp.

### CODE REVIEW

**Comment by sxjscience** [comment]
I think weight_clip is a reasonable feature and we can directly merge this PR.

**Comment by piiswrong** [comment]
please fix lint

**Comment by leezu** [comment]
I'm not quite sure why lint is not passing. For me it seems to be related to code unchanged by my commits (?).
